python-client/giskard/push/perturbation.py

```diff
@@ -0,0 +1,94 @@
+                res = Push(push_type="perturbation", feature=feat, value=ds.df.iloc[idrow][feat],
+                           bounds=bounds,
+                           perturbation_value=perturbation_res.perturbation_value)
+                yield res
+
+            # if coltype == "text" and _perturb_and_predict(model, ds, idrow, feat, coltype):
+            #     bounds = slice_bounds(feature=feat, value=ds.df.iloc[idrow][feat], ds=ds)
+            #     res = Push(push_type="perturbation", feature=feat, value=ds.df.iloc[idrow][feat],
+            #                bounds=bounds,
+            #                perturbation_value=perturbation_res.perturbation_value)
+
+            # {
+            #     "sentence": "A small variation of the text makes the prediction change, do you want to "
+            #                 "check if this unrobust behavior generalizes to the whole dataset ?",
+            #     "action": "Test",
+            #     "value": str(ds.df.iloc[idrow][feat]),
+            #     "key": str(feat),
+            # # }
+            #
+            # yield res
+
+
+def _perturb_and_predict(model, ds, idrow, feature, coltype):  # done at each step
+    def mad(x):
+        med = np.median(x)
+        x = abs(x - med)
+        mad = np.median(x)
+        return mad
+
+    ref_row = ds.df.iloc[[idrow]]
+    row_perturbed = ref_row.copy()
+    perturbation_val = None
+
+    perturbations = []
+    if coltype == "numeric":
+        mad = mad(ds.df[feature])
+        row_perturbed[feature] = _num_perturb(row_perturbed[feature], mad)
+        perturbation_val = mad
+        perturbations.append(_generate_perturbation(model, ds, ref_row, row_perturbed, perturbation_val))
+    elif coltype == "text":
+        for text_transformation in text_transfo_list:
+            ds_slice = ds.slice(lambda df: df.loc[idrow], row_level=False)
+            row_perturbed[feature] = _text_perturb(text_transformation, feature, ds_slice)
+            perturbation_val = None  # @TODO: Add text support
+            perturbations.append(_generate_perturbation(model, ds, ref_row, row_perturbed, perturbation_val))
+    return perturbations
+
+
+def _generate_perturbation(model, ds, ref_row, row_perturbed, perturbation_val):
+    if model.meta.model_type == SupportedModelTypes.CLASSIFICATION:
+        ref_prob = model.model.predict(ref_row)
+        probabilities = model.model.predict(row_perturbed)  # .reshape(1, -1)
+        return Perturbation(passed=ref_prob[0] != probabilities[0], perturbation_value=perturbation_val)
+    elif model.meta.model_type == SupportedModelTypes.REGRESSION:
+        ref_val = model.model.predict(ref_row.drop(columns=[ds.target]))
+        new_val = model.model.predict(row_perturbed.drop(columns=[ds.target]))  # .reshape(1, -1)
+        return Perturbation(passed=(new_val - ref_val) / ref_val >= 0.2, perturbation_value=perturbation_val)
+
+
+def _num_perturb(val, mad):
+    return val + 3 * mad  # 1.2  # 20% perturbation
```

Comment:
This is roughly 2 σ for a normal distribution. To me it looks like a big perturbation! Are we sure we want this? E.g. if value is close to the mean, we would be pushing it to the 95% percentile.

Reply:
Also: we only increase the value! Perturbation could be in the negative direction. (Maybe I am missing something?)